File: 4day/4daya.cpp
#include <iostream>
#include <cmath>
#include <fstream>
#include <set>
#include <string>
#include <sstream>

int main(){
    int totalPoints = 0;
    std::ifstream fin; fin.open("input.txt");
    std::set<int> winNums; std::set<int> myNums;
    std::string line;
    if(!fin.is_open()) return 1;
    while(std::getline(fin, line)){
        std::string card = line;
        //std::cout << card << std::endl;
        size_t pipe = card.find('|');
        // get #'s before pipe
        std::string beforePipe = card.substr(0, pipe); //std::cout << beforePipe << std::endl;
        std::istringstream bP(beforePipe); std::string t;
        while(bP >> t){
            try {
                int num = std::stoi(t);
                winNums.insert(num);
            } catch (const std::invalid_argument& e){
                continue;
            }
        }
        // get #'s after the pipe 
        std::string afterPipe = card.substr(pipe + 1); //std::cout << afterPipe << std::endl;
        std::istringstream aP(afterPipe);
        while(aP >> t){
            myNums.insert(std::stoi(t));
        }
        int one = 0; int twoPow = 0;
        for(auto&n:myNums){
            if(winNums.find(n) != winNums.end()){
                if(!one) one = 1; 
                else twoPow += 1;
            }
        }
        int cardTotal = one * static_cast<int>(pow(2, twoPow));
        //std::cout << cardTotal << std:: endl;
        totalPoints += cardTotal;
        winNums.clear(); myNums.clear();
    }
    fin.close();
    // 35639 is too high
    std::cout << "Total Points " << totalPoints << std::endl;
}

File: 2023/4day/4daya.py
total_points = 0

with open("input.txt", "r") as file:
    for line in file:
        card = line.strip()
        pipe = card.find('|')

        # Get #'s before pipe
        before_pipe = card[:pipe].split()
        win_nums = set()

        # Extract winning numbers
        for num in before_pipe:
            try:
                win_nums.add(int(num))
            except ValueError:
                continue

        # Get #'s after the pipe
        after_pipe = card[pipe + 1:].split()
        my_nums = set()

        # Extract your numbers
        for num in after_pipe:
            try:
                my_nums.add(int(num))
            except ValueError:
                continue

        # Calculate points
        one = 0
        two_pow = 0

        for n in my_nums:
            if n in win_nums:
                if not one:
                    one = 1
                else:
                    two_pow += 1

        card_total = one * 2**two_pow
        total_points += card_total

print("Total Points:", total_points)
